packages/plugins/response-cache/src/redis-cache.ts

```diff
@@ -0,0 +1,127 @@
+import Redis from 'ioredis';
+import type { Cache } from './cache';
+
+export type BuildRedisEntityId = (typename: string, id: number | string) => string;
+export type BuildRedisResponseOpsKey = (responseId: string) => string;
+
+export type RedisCacheParameter = {
+  /**
+   * Redis instance
+   * @see Redis.Redis https://github.com/luin/ioredis
+   */
+  redis: Redis.Redis;
+  /**
+   * Maximum amount of items in the cache. Defaults to `Infinity`.
+   */
+  max?: number;
+  /**
+   * Customize how the cache entity id is built.
+   * By default the typename is concatenated with the id e.g. `User:1`
+   */
+  buildRedisEntityId?: BuildRedisEntityId;
+  /**
+   * Customize how the cache key that stores the operations associated with the response is built.
+   * By default `operations` is concatenated with the responseId e.g. `operations:arZm3tCKgGmpu+a5slrpSH9vjSQ=`
+   */
+  buildRedisResponseOpsKey?: BuildRedisResponseOpsKey;
+};
+
+export const createRedisCache = (params: RedisCacheParameter): Cache => {
+  const store = params.redis;
+
+  const buildRedisEntityId = params?.buildRedisEntityId ?? defaultBuildRedisEntityId;
+  const buildRedisResponseOpsKey = params?.buildRedisResponseOpsKey ?? defaultBuildRedisResponseOpsKey;
+
+  function purgeEntity(entity: string): Promise<string[]> {
+    return new Promise(function (resolve, reject) {
+      const keysToPurge: string[] = [entity];
+
+      // find the responseIds for the entity
+      store.smembers(entity).then(function (responseIds) {
+        // and purge each response since they contained the entity data
+        responseIds.forEach(responseId => {
+          keysToPurge.push(responseId);
+          keysToPurge.push(buildRedisResponseOpsKey(responseId));
+        });
+
+        // if purging an entity like Comment, then also purge Comment:1, Comment:2, etc
+        if (!entity.includes(':')) {
+          store.keys(`${entity}:*`).then(function (entityKeys) {
+            for (const entityKey of entityKeys) {
+              // and purge any reponses in each of those entity keys
+              store.smembers(entityKey).then(function (responseIds) {
+                // if purging an entity check for associated operations containing that entity
+                // and purge each response since they contained the entity data
+                responseIds.forEach(responseId => {
+                  keysToPurge.push(responseId);
+                  keysToPurge.push(buildRedisResponseOpsKey(responseId));
+                });
+              });
+            }
+
+            // then purge the entityKeys like Comment:1, Comment:2 etc
+            entityKeys.forEach(entityKey => {
+              keysToPurge.push(entityKey);
+            });
+          });
+        }
+
+        // and then purge the entity itself
+        resolve(keysToPurge);
```

Comment:
This might be buggy. The `resolve(keysToPurge)` does not wait for all the async actions above.

Reply:
Oh. That's not good -- or what I want. Could you recommend another approach?

Reply:
Changed logic in `buildEntityInvalidationsKeys` such that *I think* all paths will resolve.

Reply:
Actually, I don't think my approach is correct. Revisiting.

Reply:
Have changed to use async/await which I prefer, but can change if otherwise.